server/src/test/java/org/elasticsearch/cluster/coordination/AtomicRegisterCoordinatorTests.java

```diff
@@ -0,0 +1,710 @@
+                && isLatestAcceptedStateOutdated(localAcceptedTerm, localAcceptedVersion) == false
+                // if there's a leader that's not the local node wait, otherwise win the election immediately
+                // (use the leader node id instead of equals to take into account restarts)
+                && getLeaderForTermIfAlive.apply(localCurrentTerm).map(leader -> leader.getId().equals(localNode.getId())).orElse(true)
+                && joinVotes.containsVoteFor(localNode);
+        }
+
+        private boolean isLatestAcceptedStateOutdated(long localAcceptedTerm, long localAcceptedVersion) {
+            final var currentRegisterPersistentState = sharedStore.getLatestClusterState();
+            if (currentRegisterPersistentState == null) {
+                return false;
+            }
+            return currentRegisterPersistentState.term() > localAcceptedTerm
+                || (currentRegisterPersistentState.term() == localAcceptedTerm
+                    && currentRegisterPersistentState.version() != localAcceptedVersion);
+        }
+
+        @Override
+        public boolean isPublishQuorum(
+            CoordinationState.VoteCollection voteCollection,
+            CoordinationMetadata.VotingConfiguration lastCommittedConfiguration,
+            CoordinationMetadata.VotingConfiguration latestPublishedConfiguration
+        ) {
+            assert latestPublishedConfiguration.getNodeIds().size() == 1;
+
+            return voteCollection.isQuorum(latestPublishedConfiguration);
+        }
+
+        @Override
+        public boolean shouldJoinLeaderInTerm(long currentTerm, long targetTerm) {
+            return currentTerm < targetTerm
+                // Ensure that we can join an existing leader after reading the existing cluster state from the store
+                || (currentTerm == targetTerm && currentTerm == atomicRegisterPersistedState.getInitialTermBeforeJoiningALeader());
+        }
+
+        @Override
+        public void onNewElection(DiscoveryNode sourceNode, long proposedTerm, ClusterState latestAcceptedState) {
+            final var latestClusterState = sharedStore.getLatestClusterState();
+            if (latestClusterState != null && latestClusterState.version() > latestAcceptedState.version()) {
+                throw new CoordinationStateRejectedException("The node has an stale applied cluster state version");
+            }
+
+            final var latestAcceptedClusterUUID = latestAcceptedState.metadata().clusterUUID();
+            final var proposedNewTermOwner = new TermOwner(sourceNode, proposedTerm, latestAcceptedClusterUUID);
+            final var witness = register.claimTerm(proposedNewTermOwner);
+            if (proposedNewTermOwner != witness) {
+                if (witness.clusterUUID().equals(latestAcceptedClusterUUID) == false) {
+                    atomicRegisterPersistedState.changeAcceptedClusterUUID(witness.clusterUUID());
```

Comment:
I think this should only happen before we've committed the cluster UUID?

```suggestion
                    assert latestAcceptedState.metadata().clusterUUIDCommitted() == false;
                    atomicRegisterPersistedState.changeAcceptedClusterUUID(witness.clusterUUID());
```